isomerpages/isomercms-backend
# feat(axios): implement etag functionality

## Problem
Currently, Isomer uses a huge amount of tokens due to not caching. This leads to scaling issue, where our github token limits are quickly reached and users are unable to access the CMS.

Closes [IS-172]

## Solution
1. use an external package `axios-cache-interceptor` to utilise ETags for cache. See [here](https://www.notion.so/opengov/A-brief-primer-on-ETags-94669d1abed44dcbb02ee0ffc36d0a7d?pvs=4) for info on etags. 
    - note that this is using **in memory** storage. the implications are that: we could potentially have some requests that **could be cached** but are not (current ETag value is 1 -> user A edits -> ETag value now 2 -> other instance doesn't knows -> queries from other instance -> ETag value updated)
    - potential for stale reads - github will tell us if the old etag value is outdated or not (ie, return 200) but the package uses the `cache-control` header to control freshness (which is good). because github's header sets `max-age=60`, the package will internally use the old value for 60s. 

## Tests
1. first, add a `console.log` for `remainingRequests` into `respHandler` for `AxiosInstance.ts`, so that we know how much requests are remaining 
3. next, issue a `GET` request to the `e2e-test-repo`'s homepage (`curl --request GET \
  --url http://localhost:8081/v2/sites/e2e-test-repo/homepage`) 
4. observe the # of requests remaining. 
5. repeat steps 2-3; the # of requests remaining should be **unchanged**
6. make an edit on the [homepage](https://github.com/isomerpages/e2e-test-repo/blob/staging/index.md) 
7. repeat steps 2-3, the # of requests should be **unchanged if it's less than a minute**
8. wait a minute
9. do steps 2-3, the # of requests should be **decreased**

**New dependencies**:
- `axios-cache-interceptor` : interceptor for axios to use etags. the current backing storage is in memory.

**Deploy notes**
1. `axios` got upgraded from `0.21.x` to `0.25.0` 
    - list of breaking changes here 
    - went through most of [them](https://github.com/axios/axios/blob/main/CHANGELOG.md#0250-january-18-2022), mostly typing changes. 
2. had to use `0.9.2` as a minimum for `axios-cache-interceptor`
    - see compat chart [here](https://axios-cache-interceptor.js.org/v0/guide/getting-started)
    - this is because of 2 reasons - `ETag` functionality only added in `0.6` + their import for `object-code` is an outdated one, which doesn't get fixed till `0.9.2`. 

[IS-172]: https://isomeropengov.atlassian.net/browse/IS-172?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
